lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```diff
@@ -0,0 +1,271 @@
+  EXIT_ON_ERR(::glow::optimizeFunction(&F, backend, cctx));
+  for (const Node &N : F.getNodes()) {
+    CHECK(backend.isOpSupported(N))
+        << "Backend must support all nodes after high-level optimizations but "
+           "encountered unsupported operator: "
+        << N.getDebugDesc();
+  }
+  auto funcOrErr = backend.compile(&F, cctx.backendOpts);
+  EXIT_ON_ERR(funcOrErr.takeError());
+  return std::move(*funcOrErr);
+}
+
+/// Runs the compiled function \p compiledF on the \p backend using provided \p
+/// bindings.
+void run(Backend &backend, CompiledFunction &compiledF,
+         PlaceholderBindings &bindings) {
+  std::unique_ptr<PlaceholderBindings> bindingsPtr(&bindings);
+  ExecutionContext context(std::move(bindingsPtr));
+  // TODO: Add only constants used by F to the compiled function. This should
+  // reduce the amount of data that needs to be copied.
+  auto executeErr = compiledF.execute(&context);
+  EXIT_ON_ERR(std::move(executeErr));
+  // don't delete bindings.
+  context.movePlaceholderBindings().release();
+}
+
+/// Evaluates a provided constant operation \p C using the provided \p backend
+/// and using the compilation context \p cctx.
+/// \returns constant results.
+std::vector<Constant *>
+evaluateConstantOperation(Backend &backend, CompilationContext &cctx, Node *C) {
+  PlaceholderBindings bindings;
+  assert(isConstantOperation(C, backend) && "Expected a constant expression");
+  // Constants and splats do not need to be constant evaluated.
+  if (isa<Constant>(C) || isa<SplatNode>(C)) {
+    return {};
+  }
+  Module &mod = *C->getParent()->getParent();
+  // Create a temporary function to perform the constant operation.
+  Function *constEvaluationF = mod.createFunction(constEvaluationFunctionName);
+  // Mapping from existing nodes to the new ones.
+  NodeMap currToNew;
+  // Clone the constant operation and some of its inputs if necessary.
+  auto *clonedC = recursiveClone(constEvaluationF, C, currToNew);
+  // Create save nodes for each of the results.
+  llvm::SmallVector<SaveNode *, 16> savedResults;
+  for (size_t idx = 0, e = clonedC->getNumResults(); idx < e; ++idx) {
+    auto *SN = constEvaluationF->createSave(clonedC->getName(),
+                                            clonedC->getNthResult(idx));
+    savedResults.emplace_back(SN);
+    bindings.allocate(SN->getPlaceholder());
+  }
+  // Run the temporary backend to perform this constant operation
+  // evaluation.
+  EXIT_ON_ERR(
+      executeConstantFunction(backend, *constEvaluationF, bindings, cctx));
+  // Get the results of the constant operation compile-time computation and
+  // create new constants from it.
+  std::vector<Constant *> constResults;
+  constResults.reserve(savedResults.size());
+  for (auto *SN : savedResults) {
+    Tensor *outputTensor = bindings.get(SN->getPlaceholder());
+    auto *constResult = mod.createConstant(SN->getName(), *outputTensor);
```

Comment:
nit: `std::move(*outputTensor)` here so we don't need to copy the tensor payload.